core/src/main/scala/org/bykn/bosatsu/Evaluation.scala

```diff
@@ -150,22 +258,22 @@ case class Evaluation(pm: PackageMap.Inferred, externals: Externals) {
          val earg = recurse((p, Right(arg), env))._1
          (for {
            fn <- efn
-           afn = fn.asInstanceOf[Fn[Any, Any]] // safe because we typecheck
+           afn = fn.asInstanceOf[FnValue] // safe because we typecheck
            a <- earg
-         } yield afn(a), resT)
+           res <- afn.toFn(a)
+         } yield res, resT)
        case a@AnnotatedLambda(name, argt, expr, _) =>
-         val fn = new Fn[Any, Any] {
-           override def toString = a.repr
-           def apply(x: Any) =
-             recurse((p, Right(expr), env + (name -> x)))._1.value
-         }
+         val fn =
+           FnValue { x =>
+             recurse((p, Right(expr), env + (name -> x)))._1
```

Comment:
if we separate this function do `(Package, Ref) => Map[String, Eval[Value]] => Eval[Value]` we can run the outer function before we get `x` then do less recomputation on each `x` value.